teps/0061-allow-custom-task-to-be-embedded-in-pipeline.md

```diff
@@ -0,0 +1,277 @@
+See issue [tektoncd/pipeline#3682](https://github.com/tektoncd/pipeline/issues/3682)
+
+### Goals
+
+1. Allow custom tasks to be embedded in a pipeline specification,
+2. Custom task specification verification/validation should be handed
+over to custom task controllers, custom task specification must not be
+validated by Tektoncd/Pipeline validation logic.
+3. Custom taskSpec should be submitted as part of the runSpec.
+
+### Non-Goals
+
+Custom task controllers are to be developed by other parties. Custom task
+specification should not be validated by Tektoncd/Pipeline webhooks.
+
+### Use Cases (optional)
+
+<!--
+Describe the concrete improvement specific groups of users will see if the
+Motivations in this doc result in a fix or feature.
+
+Consider both the user's role (are they a Task author? Catalog Task user?
+Cluster Admin? etc...) and experience (what workflows or actions are enhanced
+if this problem is solved?).
+-->
+
+- Users can put all the information in one pipelineRun CR.
+- Users don't have to manage the custom task CR. Since custom task CR is
+namespace scope. Currently if there are multiple users in the same namespace, they
+will have conflicts when they are using the same name for their custom task CR.
+- In KFP, we need all the templates and task spec live in each pipeline. Currently, 
+having all the custom task templates living in the Kubernetes namespace scope means that
+we have to make multiple API calls to Kubernetes in order to get all the pipeline
+information to render in our API/UI. 
+
+
+## Requirements
+
+<!--
+Describe constraints on the solution that must be met. Examples might include
+performance characteristics that must be met, specific edge cases that must
+be handled, or user scenarios that will be affected and must be accomodated.
+-->
+
+- The Tekton controller is responsible for adding the the custom task spec to
+the Run spec. Then it get validate by the custom task controllers.
+
+## Proposal
+
+Use combination of taskRef and taskSpec to indicate an embedded custom task.
+Currently taskRef and taskSpec are mutually exclusive. This behavior will
+have to be changed. For example, a pipeline spec may look like this:
+
+```
+spec:
+  tasks:
+    - name: run-my-custom-task
+      taskRef:
+        apiVersion: example.tekton.dev/v1alpha1
+        kind: Example
+      taskSpec:
+        key1: value1
+        key2:
+          keykey1: value11
+        stage1:
+          dothis: ok
+        stage2:
+          dothat: ok
+```
+
+Use Tektoncd/pipeline webhook to determine if a task is a custom task based
+on above suggestion. Once a task is determined to be a custom task, then the
+specification defined in taskSpec will be also used to define the `Run` custom
+resource. This requires the `Run` API also need to be updated with the taskSpec.
+
+Once `Run` resource creation is requested by Tektoncd/Pipeline webhook,
+the custom controller should have a webhook to validate if the `Run` has the embedded
```

Comment:
I don't think we should prescribe that a Run controller _should_ use webhook validation necessarily (even though I think it's a good idea). Webhooks in k8s are unfortunately still very annoying to set up. Async validation in reconciling might be better for some Custom Tasks, so we should allow that.

Maybe loosen the wording to "custom controllers should consider having a webhook to validate..." ?

Reply:
Thanks, the webhook could be optional and let the custom task owner decide whether to implement it.